runtime/gc_ctrl.c

```diff
@@ -103,20 +103,17 @@ CAMLprim value caml_gc_minor_words(value v)
 
 CAMLprim value caml_gc_counters(value v)
 {
-  CAMLparam0 ();   /* v is ignored */
-  CAMLlocal1 (res);
-
   /* get a copy of these before allocating anything... */
   double minwords = caml_gc_minor_words_unboxed();
   double prowords = (double)Caml_state->stat_promoted_words;
   double majwords = Caml_state->stat_major_words +
                     (double) Caml_state->allocated_words;
 
-  res = caml_alloc_3(0,
-    caml_copy_double (minwords),
-    caml_copy_double (prowords),
-    caml_copy_double (majwords));
-  CAMLreturn (res);
+  v = caml_alloc_small(3 * Double_wosize, Double_array_tag);
+  Store_double_flat_field(v, 0, minwords);
+  Store_double_flat_field(v, 1, prowords);
+  Store_double_flat_field(v, 2, majwords);
+  return v;
```

Comment:
I don't think that this is correct, because you are changing the return type without changing the OCaml consumers. It would be much simpler (in terms of reasoning about backward compatibility) to keep returning a tuple of boxed floats.